pkg/registry/certificates/clustertrustbundle/strategy.go

```diff
@@ -0,0 +1,147 @@
+var _ rest.RESTDeleteStrategy = Strategy
+
+func (strategy) NamespaceScoped() bool {
+	return false
+}
+
+func (strategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {}
+
+func noRestrictionsOnName(name string, prefix bool) []string {
+	return nil
+}
+
+func (strategy) Validate(ctx context.Context, obj runtime.Object) field.ErrorList {
+	bundle := obj.(*certificates.ClusterTrustBundle)
+
+	var allErrors field.ErrorList
+	allErrors = append(allErrors, apivalidation.ValidateObjectMeta(&bundle.ObjectMeta, false, noRestrictionsOnName, field.NewPath("metadata"))...)
+
+	if bundle.Spec.SignerName != "" {
+		signerNameErrors := certvalidation.ValidateSignerName(field.NewPath("spec", "signerName"), bundle.Spec.SignerName)
+		allErrors = append(allErrors, signerNameErrors...)
+	}
+
+	// TODO(KEP-3257): Is it OK to modify the object during validate?
+	pemTrustAnchors, err := normalizePEMTrustAnchors(bundle.Spec.PEMTrustAnchors)
+	if err != nil {
+		allErrors = append(allErrors, field.Invalid(field.NewPath("spec", "pemTrustAnchors"), "<pemTrustAnchors after normalization>", "contains an invalid block"))
+		return allErrors
+	}
+	bundle.Spec.PEMTrustAnchors = pemTrustAnchors
+
+	if bundle.Spec.PEMTrustAnchors == "" {
+		allErrors = append(allErrors, field.Invalid(field.NewPath("spec", "pemTrustAnchors"), "<pemTrustAnchors after normalization>", "at least one trust anchor must be provided"))
+	}
+
+	return allErrors
+}
+
+func (strategy) WarningsOnCreate(ctx context.Context, obj runtime.Object) []string {
+	return nil
+}
+
+func (strategy) Canonicalize(obj runtime.Object) {
+	// TODO(KEP-3257): Is it OK to do the canonicalization as part of
+	// Validate(), or do we need an explicit call to normalizePEMTrustAnchors
+	// here?
+}
+
+func (strategy) AllowCreateOnUpdate() bool {
+	return false
+}
+
+func (s strategy) PrepareForUpdate(ctx context.Context, new, old runtime.Object) {
+	s.PrepareForCreate(ctx, new)
+}
+
+func (s strategy) ValidateUpdate(ctx context.Context, new, old runtime.Object) field.ErrorList {
+	oldBundle := old.(*certificates.ClusterTrustBundle)
+	newBundle := new.(*certificates.ClusterTrustBundle)
+
+	var allErrors field.ErrorList
+	allErrors = append(allErrors, s.Validate(ctx, newBundle)...)
+	allErrors = append(allErrors, apivalidation.ValidateObjectMetaUpdate(&newBundle.ObjectMeta, &oldBundle.ObjectMeta, field.NewPath("metadata"))...)
+
+	if newBundle.Spec.SignerName != oldBundle.Spec.SignerName {
+		allErrors = append(allErrors, field.Forbidden(field.NewPath("spec", "signerName"), "updates may not change the signer name"))
+	}
+
+	return allErrors
+}
+
+func (strategy) WarningsOnUpdate(ctx context.Context, obj, old runtime.Object) []string {
+	return nil
+}
+
+func (strategy) AllowUnconditionalUpdate() bool {
+	return true
+}
+
+// normalizePEMTrustAnchors strips interblock data, strips in-block headers,
+// reserializes the blocks to remove line-wrapping and padding differences, then
+// sorts them alphabetically.
```

Comment:
Basically to prevent clients from depending on having a specific order available.  It would be even better if we could find a way to make the order stable but unpredictable (maybe by keying the sort order on the overall hash of all the certificates)?

In-block headers are cut because we re-encode the PEM blocks to remove linewrapping differences (and also because in-block headers are a bad idea).

Interblock comments are cut as a consequence of sorting the certs (and also because they're not a great idea, people will be tempted to use them for signaling).

Reply:
I do not think we should mutate the user's input in any way.  If the input is valid, it should be persisted as-is.  It makes the API contract nicer for humans.

However, as far as code consuming this is concerned:

Randomizing order is good.
Normalizing line wrapping is good.
Removing comments is good.

All of that should be done by kubelet before it mounts stuff into pods.  If direct consumers of the REST API want to be bad, they could simply use labels and annotations to do so.

Reply:
Hmmm, I didn't think about having Kubelet do this.  You're right, that's a better idea.

Reply:
I think this is done.  We no longer mutate user input during validation.